### PR TITLE
chore: Add dtslint and Cypress static types

### LINKED ISSUE #1046: Refactor TypeScript types and add types for Cypress._, Cypress.$ etc
- Refactor and update `cli/index.d.ts` TS types
- Add dtslint tests
- Add types to Cypress wrappings like Cypress.$, Cypress._, etc

### PR DESCRIPTION
**Pull request by NicholasBoll**
This is a break-apart of more completed work. Info can be found at #1040 

This particular PR closes #1046 

Included in this PR:
* Add https://github.com/Microsoft/dtslint for type tests
* Move types into a folder called `types` per `dtslint` recommendations (there are a few types files)
* Add type definitions for `Cypress` aliased objects like `Cypress._`, `Cypress.$`, etc.
* Add `dtslint` to unit test task. Not sure the most appropriate spot

I didn't actually add any type tests as all my tests are verifying the completed type definitions.

### CODE REVIEW

File: cli/package.json [hunk 1]
```diff
@@ -11,20 +11,28 @@
...
     "build": "node ./scripts/build.js",
     "prerelease": "npm run build",
     "release": "cd build && releaser --no-node --no-changelog",
     "size": "t=\"$(npm pack .)\"; wc -c \"${t}\"; tar tvf \"${t}\"; rm \"${t}\";"
   },
+  "types": "types",
   "dependencies": {
     "@cypress/listr-verbose-renderer": "0.4.1",
     "@cypress/xvfb": "1.0.4",
+    "@types/blob-util": "1.3.3",
```
Comment (NicholasBoll) [inline cli/package.json]: I'm pretty sure these have to be normal dependencies for downstream projects. I haven't actually deployed and linked as `devDependencies`, but from other testing the types will fail and fallback to `any` if the types can't be found.

Comment (bahmutov) [reply cli/package.json]: yeah, I think they will have to be included then.

File: cli/types/index.d.ts [hunk 2]
```diff
@@ -910,6 +927,165 @@ declare namespace Cypress {
...
+     * @default 60000
+     */
+    pageLoadTimeout: number;
+    /**
+     * Time, in milliseconds, to wait for an XHR request to go out in a [cy.wait()](https://on.cypress.io/api/wait) command
+     * @default 5000
+     */
+    requestTimeout: number;
+    /**
+     * Time, in milliseconds, to wait until a response in a [cy.request()](https://on.cypress.io/api/request), [cy.wait()](https://on.cypress.io/api/wait), [cy.fixture()](https://on.cypress.io/api/fixture), [cy.getCookie()](https://on.cypress.io/api/getcookie), [cy.getCookies()](https://on.cypress.io/api/getcookies), [cy.setCookie()](https://on.cypress.io/api/setcookie), [cy.clearCookie()](https://on.cypress.io/api/clearcookie), [cy.clearCookies()](https://on.cypress.io/api/clearcookies), and [cy.screenshot()](https://on.cypress.io/api/screenshot) commands
```
Comment (bahmutov) [inline cli/types/index.d.ts]: you don't have to use `https://on.cypress.io/api/request` just `https://on.cypress.io/request` would be better!

File: cli/types/index.d.ts [hunk 3]
```diff
@@ -910,6 +927,165 @@ declare namespace Cypress {
     cancable?: boolean;
   }
 
+  interface ConfigOptions {
```
Comment (bahmutov) [inline cli/types/index.d.ts]: so we allow all options to be optional using `Partial<ConfigOptions>`? that is pretty slick if true!

**Comment by bahmutov** [comment]
@NicholasBoll do you see the broken build in https://circleci.com/gh/cypress-io/cypress/7682 ? since you moved `cli/index.d.ts` which we included in the built `cypress` package - do we have to copy `cli/types` to `cli/build/types` during `npm run build` in `cli` folder as well?  Because we are building "production" `cypress` NPM package 

```
cd cli
npm run build
```

File: cli/types/index.d.ts (hunk 3 above)
Comment (NicholasBoll) [reply cli/types/index.d.ts]: Yes. `Partial` is a mapped type that iterates over the keys and makes each optional. It is really useful if you want an interface without optional parameters and make them optional for things like default overrides.

Mapped types are pretty cool. TypeScript `lib.d.ts` defines a few globally: https://github.com/Microsoft/TypeScript/blob/master/lib/lib.d.ts#L1329

File: cli/types/index.d.ts (hunk 2 above)
Comment (NicholasBoll) [reply cli/types/index.d.ts]: Sure thing.

There are a lot of other URLs with the extra `/api` in the URL. I can update those URLs as well.

**Comment by NicholasBoll** [comment]
I'll fix the build issue. I'll use #979 to make sure get all the path changes corrected. Types aren't required to go into a `types` folder, but I think it helps reduce the clutter in the `cli` directory since there are supporting files and type tests.

**Comment by NicholasBoll** [comment]
I'll also make sure the package contains all the type definitions using `npm pack`.

**Comment by NicholasBoll** [comment]
I've noticed that JS files don't have semicolons. I think the TSLint and type definitions shouldn't have them either.

**Comment by bahmutov** [comment]
all great stuff @NicholasBoll ! hope to merge it

**Comment by NicholasBoll** [comment]
I've removed semicolons from the type definitions files.

Removing semicolons from `index.d.ts` was enough to not trigger git's threshold on moved files. So now the PR shows removing `cli/index.d.ts` and adding `cli/types/index.d.ts`

@bahmutov You'll have to let me know if that's okay. If not, I can remove the semicolon removal as part of a separate PR.

**Comment by NicholasBoll** [comment]
I've added type tests: `types/tests/cypress-tests.ts`. Running `npm run dtslint` from inside the `cli` directory will validate those tests pass.

**Comment by bahmutov** [comment]
Love it, locally checks out!

**Comment by brian-mann** [comment]
Fixes #1046

**Pull request merged**